File: ants.py
import numpy as np
import math as mt
import random
import pygame
from pygame import gfxdraw
import sys

COLOR_red = (255,0,0)
COLOR_white = (255,255,255)
COLOR_black = (0,0,0)

class ant:
	def __init__(self, position, loaded, field_of_view, index_in_position_array):
		self.position = position
		self.loaded = loaded
		self.field_of_view = field_of_view
		self.index_in_position_array = index_in_position_array

	def pick_up_ant(self, board):
		board[self.position] = 0
		self.loaded = True

	def drop_off_ant(self, board):
		board[self.position] = -1
		self.loaded = False

	def get_field_of_view_coords(self, board):
		board_lines = board.shape[0]
		board_columns = board.shape[1]
		fov_coords = []
		for i in range(-self.field_of_view, self.field_of_view +1):
			for j in range(-self.field_of_view, self.field_of_view +1):
				fov_coords.append(((self.position[0]+i)%board_lines, (self.position[1]+j)%board_columns))
		return fov_coords

	def should_pick_up(self, board):
		fov_coords = self.get_field_of_view_coords(board)
		return random.random() <= (float(len([board[k] for k in fov_coords if board[k] == 0]))/len(fov_coords))

	def should_drop_off(self, board):
		fov_coords = self.get_field_of_view_coords(board)
		return random.random() <= (float(len([board[k] for k in fov_coords if board[k] == -1]))/len(fov_coords))

	def move(self, board, live_ants_positions):
		# |0|1|2|
		# |3|x|4|
		# |5|6|7|
		board_lines = board.shape[0]
		board_columns = board.shape[1]
		coords = {0: (-1, -1), 1: (-1, 0), 2: (-1, 1), 3: (0, -1), 4: (0, 1), 5: (1, -1), 6: (1, 0), 7: (1, 1)}
		direction = random.randint(0,7)
		new_position = ((self.position[0] + coords[direction][0])%board_lines, (self.position[1] + coords[direction][1])%board_columns)
		old_position = (self.position[0], self.position[1])
		#if not new_position in live_ants_positions:
			#live_ants_positions[live_ants_positions.index(self.position)] = new_position
		live_ants_positions[self.index_in_position_array] = new_position
		self.position = new_position
		return old_position, new_position
		

def generate_board(lines, columns):
	return np.zeros(shape=(lines, columns))

def populate_board(board, n_dead_ants):
	board_columns = board.shape[1]
	board_size = board.size
	dead_ants_positions = random.sample(range(board_size), n_dead_ants)
	dead_ants_coord = []
	for i in dead_ants_positions:	dead_ants_coord.append((i/board_columns, i%board_columns))
	for i in dead_ants_coord:	board[i]= -1

def generate_live_ants(board, fov_range, n_live_ants):
	live_ants = []
	index = 0
	for i in random.sample(range(board.size), n_live_ants):
		live_ants.append(ant((i/board.shape[1], i%board.shape[1]), False, fov_range, index))
		index += 1
	return live_ants
	#return [ant(k, False, 1) for k in [(i/board.shape[1], i%board.shape[1]) for i in random.sample(range(board.size), n_live_ants)]]

def check_closed_window():
	for event in pygame.event.get():
	  if event.type == pygame.QUIT:
	    pygame.quit(); sys.exit();


pygame.init()
screen = pygame.display.set_mode((500,500))
screen.fill(COLOR_white)

pygame.display.update()

board = generate_board(500,500)
populate_board(board, 5000)
ants = generate_live_ants(board, 3, 500)
live_ants_positions = [ant.position for ant in ants]

for i in range(500):
	for j in range(500):
		if board[i, j] == -1:
			gfxdraw.pixel(screen, i, j, COLOR_red)
		if (i, j) in live_ants_positions:
			gfxdraw.pixel(screen, i, j, COLOR_black)
			
pygame.display.update()
i=1
while True:
	check_closed_window()
	for ant in ants:
		if not ant.loaded:
			if board[ant.position] == -1 and ant.should_pick_up(board):
				ant.pick_up_ant(board)
		elif ant.loaded:
			if board[ant.position] == 0 and ant.should_drop_off(board):
				ant.drop_off_ant(board)
		
		old_position, new_position = ant.move(board, live_ants_positions)
		
		if board[old_position] == -1:
			gfxdraw.pixel(screen, old_position[0], old_position[1], COLOR_red)
		else:
			gfxdraw.pixel(screen, old_position[0], old_position[1], COLOR_white)
			
		gfxdraw.pixel(screen, new_position[0], new_position[1], COLOR_black)
		
	i += 1
	if i%1000 == 0:
		pygame.display.update()

File: ants.cpp
//compile: g++ -std=c++11 ants.cpp -o ants -lsfml-graphics -lsfml-window -lsfml-system

#include <iostream>
#include <tuple>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <cstring>
#include <algorithm>
//#include <SFML/Graphics.hpp>

using namespace std;

typedef struct
{
  int old_i, old_j, new_i, new_j;
}Position;

pair<int, int> coords[8] = {  pair<int, int>(-1, -1), pair<int, int>(-1, 0), pair<int, int>(-1, 1), pair<int, int>(0, -1), 
                              pair<int, int>(0, 1), pair<int, int>(1, -1), pair<int, int>(1, 0), pair<int, int>(1, 1)
                            };
int mod (int a, int b)
{
  if(b < 0)
    return mod(a, -b);   
  int ret = a % b;
  if(ret < 0)
    ret+=b;
  return ret;
}

class Ant 
{
  public:
    pair<int, int> position;
    bool loaded;
    int field_of_view;
    int index_in_position_array;
    Ant(pair<int, int> position, bool loaded, int field_of_view, int index_in_position_array);
    void pick_up_ant(vector<vector<int> > & board);
    void drop_off_ant(vector<vector<int> > & board);
    //vector<pair<int,int> > get_field_of_view_coords(vector<vector<int> > board);
    bool should_pick_up(vector<vector<int> > board);
    bool should_drop_off(vector<vector<int> > board);
    //int move(vector<vector<int> > board, vector<pair<int,int> > & live_ants_positions, vector<vector<int> > live_ants_board, pair<pair<int, int>, pair<int, int> > & positions);
    int move(vector<vector<int> > board, vector<vector<int> > live_ants_board, Position* positions);
};

Ant::Ant(pair<int, int> positio, bool loade, int field_of_vie, int index_in_position_arra)
{
  position = positio;
  loaded = loade;
  field_of_view = field_of_vie;
  index_in_position_array = index_in_position_arra;
}

void Ant::pick_up_ant(vector<vector<int> > & board)
{
  board[this->position.first][this->position.second] = 0;
  this->loaded = true;
}

void Ant::drop_off_ant(vector<vector<int> > & board)
{
  board[this->position.first][this->position.second] = -1;
  this->loaded = false;
}

bool Ant::should_pick_up(vector<vector<int> > board)
{
  // vector<pair<int, int> > fov_coords = this->get_field_of_view_coords(board);
  // int cont = 0;
  // for (vector<pair<int, int> >::iterator it = fov_coords.begin(); it != fov_coords.end(); ++it)
  // {
  //  if (board[it->first][it->second] == 0)
  //  {
  //    cont++;
  //  }
  // }
 //  return (((rand()%100)/100) <= cont/fov_coords.size());
  int board_lines = board.size();
  int board_columns = board[0].size();
  int cont=0;
  int total =0;
  for (int i = -this->field_of_view; i <= this->field_of_view; ++i)
  {total++;
    for (int j = -this->field_of_view; j <= this->field_of_view; ++j)
    {
      if (board[mod(this->position.first + i, board_lines)][mod(this->position.second + j, board_columns)] == 0)
        cont++;
    }
  }

  return (((random()%100)/100) <= cont/total);
}

bool Ant::should_drop_off(vector<vector<int> > board){
  // //vector<pair<int, int> > fov_coords = this->get_field_of_view_coords(board);
  // int cont = 0;
  // for (vector<pair<int, int> >::iterator it = fov_coords.begin(); it != fov_coords.end(); ++it)
  // {
  //   if (board[it->first][it->second] == -1)
  //   {
  //     cont++;
  //   }
  // }
  int board_lines = board.size();
  int board_columns = board[0].size();
  int cont=0;
  int total =0;
  for (int i = -this->field_of_view; i <= this->field_of_view; ++i)
  {total++;
    for (int j = -this->field_of_view; j <= this->field_of_view; ++j)
    {
      if (board[mod(this->position.first + i, board_lines)][mod(this->position.second + j, board_columns)] == -1)
        cont++;
    }
  }

  return (((random()%100)/100) <= cont/total);
}

// vector<pair<int,int> > Ant::get_field_of_view_coords(vector<vector<int> > board)
// {
//   int board_lines = board.size();
//   int board_columns = board[0].size();
//   vector<pair<int,int> > fov_coords;
//   for (int i = -this->field_of_view; i <= this->field_of_view; ++i)
//   {
//     for (int j = -this->field_of_view; j <= this->field_of_view; ++j)
//     {
//       fov_coords.push_back(pair<int, int>(mod(this->position.first + i, board_lines), mod(this->position.second + j, board_columns)));
//     }
//   }
//   return fov_coords;
// }

int Ant::move(vector<vector<int> > board, vector<vector<int> > live_ants_board, Position* positions)
{
    //  |0|1|2|
    //  |3|x|4|
    //  |5|6|7|
  int board_lines = board.size();
  int board_columns = board[0].size();
  int direction = random()%8;
  if (live_ants_board[mod(this->position.first + coords[direction].first, board_lines)][mod(this->position.second + coords[direction].second, board_columns)] == 1)
  {
    positions->new_i = this->position.first;
    positions->new_j = this->position.second;
    positions->old_i = this->position.first;
    positions->old_j = this->position.second;
    return 0;
  }
  positions->new_i = mod(this->position.first + coords[direction].first, board_lines);
  positions->new_j = mod(this->position.second + coords[direction].second, board_columns);
  positions->old_i = this->position.first;
  positions->old_j = this->position.second;
  this->position.first = positions->new_i;
  this->position.second = positions->new_j;
  
  return 1;
}

vector<vector<int> > generate_board(int lines, int columns)
{
  vector<vector<int> > board(lines);
  for ( int i = 0 ; i < lines ; i++ )
    board[i].resize(columns);
  return board;
}

void populate_board(vector<vector<int> > & board, int n_dead_ants)
{
  int board_lines = board.size();
  int board_columns = board[0].size();
  vector<int> total_positions;
  for (int i = 0; i < board_lines*board_columns; ++i)
    total_positions.push_back(i);
  random_shuffle(total_positions.begin(), total_positions.end());
  for (int i = 0; i < n_dead_ants; ++i)
    board[total_positions[i]/board_columns][total_positions[i]%board_columns] = -1;
  return;
}

void populate_live_ants_board(vector<vector<int> > & board, vector<Ant> live_ants)
{
  for (vector<Ant>::iterator i = live_ants.begin(); i != live_ants.end(); ++i)
  {
    board[i->position.first][i->position.second] = 1;
  }
}

vector<Ant> generate_live_ants(vector<vector<int> > board, int fov_range, int n_live_ants)
{
  vector<Ant> live_ants;
  int board_lines = board.size();
  int board_columns = board[0].size();
  vector<int> total_positions;
  for (int i = 0; i < board_lines*board_columns; ++i)
    total_positions.push_back(i);
  random_shuffle(total_positions.begin(), total_positions.end());
  for (int i = 0; i < n_live_ants; ++i)
  {
    Ant ant(pair<int, int>(total_positions[i]/board_columns,total_positions[i]%board_columns), false, fov_range, i); 
    live_ants.push_back(ant);
  }
  return live_ants;
}

// void show_board(vector<vector<int> > board)
// {
//   for (int i = 0; i < board.size(); ++i)
//   {  
//     for (int j = 0; j < board[0].size(); ++j)
//     {
//       cout << board[i][j] << '\t';
//     }
//     cout << '\n';
//   }
//   return;
// }

// void draw_board(sf::RenderWindow & windowRef, vector<vector<int> > board, vector<vector<int> > live_ants_board)
// {
//   int board_lines = board.size();
//   int board_columns = board[0].size();
//   windowRef.clear(sf::Color::White);
//   for (int i = 0; i < board_lines; ++i)
//   {
//     for (int j = 0; j < board_columns; ++j)
//     {
//       sf::RectangleShape rectangle;
//       rectangle.setSize(sf::Vector2f(1, 1));
//       rectangle.setPosition(i, j);
//       if(board[i][j]==-1)
//       {
//         rectangle.setFillColor(sf::Color::Red);
//         windowRef.draw(rectangle);
//       }
//       if (live_ants_board[i][j]==1)
//       {
//         rectangle.setFillColor(sf::Color::Black);
//         windowRef.draw(rectangle);
//       }
      
//     }
//   }
// }

int main () {
  
  srand(time(NULL));

  //sf::RenderWindow window(sf::VideoMode(160, 160), "Ants");

  vector<vector<int> > board = generate_board(160,160);
  vector<vector<int> > live_ants_board = board;
  populate_board(board, 7000);

  vector<Ant> live_ants = generate_live_ants(board, 3, 100);
  populate_live_ants_board(live_ants_board, live_ants);
  //pair<pair<int, int>, pair<int, int> > positions;
  Position positions;
  int i = 0;
  while(i<=2000)
  {
    // if (i%100==0)
    // {
    //   draw_board(window, board, live_ants_board);
    //   window.display();
    // }
    // sf::Event event;
    // while (window.pollEvent(event))
    // {
    //   // "close requested" event: we close the window
    //   if (event.type == sf::Event::Closed)
    //   {
    //     window.close();
    //     return 0;
    //   }
    // }
    //cout << i << '\n';
    for (vector<Ant>::iterator ant = live_ants.begin(); ant != live_ants.end(); ++ant)
    {
      if (!ant->loaded)
      {
        if (board[ant->position.first][ant->position.second] == -1 && ant->should_pick_up(board))
        {
          ant->pick_up_ant(board);
        }
      } 
      else if (ant->loaded)
      {
        if (board[ant->position.first][ant->position.second] == 0 && ant->should_drop_off(board))
        {
          ant->drop_off_ant(board);
        }
      }
      ant->move(board, live_ants_board, &positions);
      live_ants_board[positions.old_i][positions.old_j] = 0;
      live_ants_board[positions.new_i][positions.new_j] = 1;
    }
    i++;
  }

  return 0;
}
